### PR TITLE
Panns

### PR DESCRIPTION
**Pull request by aryamansriram**
This branch consists of the following:
a. models/panns_finetune.py: Finetuning script for the bird panns
b. model_configs/panns: Configs for the panns scripts
c. predictions/panns: Contains scripts for inference and getting the metrics

### CODE REVIEW

**Review by nandan-sridhar — COMMENTED** [review]
Please add a README

File: predictions/panns/panns_kaggle1_inference.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+                            detected[tail_idx + 1] -  # noqa W504
+                            detected[tail_idx] != 1):
+                        onset = 0.01 * detected[
+                            head_idx] + global_time
+                        offset = 0.01 * detected[
+                            tail_idx] + global_time
+                        onset_idx = detected[head_idx]
+                        offset_idx = detected[tail_idx]
+                        max_confidence = framewise_outputs[
+                                         onset_idx:offset_idx, target_idx].max()  # noqa E126
```
Comment (nandan-sridhar) [inline predictions/panns/panns_kaggle1_inference.py:108]: Please don't suppress any of these warnings. Use backslash for continuations to the next line

File: predictions/panns/panns_kaggle1_inference.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+            predictions["ebird_code"].append(clip_codes)
+        else:
+            predictions["ebird_code"].append("NOCALL")
+
+    return pd.DataFrame(predictions)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Makes Predictions using the Panns Model")
+    parser.add_argument("-p", "--path_to_input", action="store", required=True, help="Path to the input data")
```
Comment (nandan-sridhar) [inline predictions/panns/panns_kaggle1_inference.py:177]:
```suggestion
    parser.add_argument("-p", "--path_to_input", action="store", required=True, help="Path to csv file with input manifest")
```

File: predictions/panns/panns_kaggle1_inference.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        # print("Iteration ",ii+1,"of ",len(paths))
+        try:
+            clip, _ = librosa.load(paths[ii], sr=SR, mono=True, res_type="kaiser_fast")
+            total += 1
+        except:
+            print("Something wrong with the audioclip")
+
+            continue
+
+        pred_df, clip_codes = prediction_for_clip(clip, model, model_details["threshold"],
```
Comment (nandan-sridhar) [inline predictions/panns/panns_kaggle1_inference.py:157]: Can we get probabilities for each label?

File: predictions/panns/panns_kaggle1_inference.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
+# (https://www.apache.org/licenses/LICENSE-2.0)
```
[the remark this answers is not in the document]
Comment (aryamansriram) [reply predictions/panns/panns_kaggle1_inference.py:1]: Done

File: models/panns_xeno_finetune.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
+import argparse
```
[the remark this answers is not in the document]
Comment (aryamansriram) [reply models/panns_xeno_finetune.py:1]: Done

File: models/panns_xeno_finetune.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+            print(e)
+            inp = input("Enter y to break")
+            if inp == "y":
+                break
+            print(new_df_dict)
+
+    return pd.DataFrame(new_df_dict)
+
+
+class XC_Dataset(Dataset):
```
[the remark this answers is not in the document]
Comment (aryamansriram) [reply models/panns_xeno_finetune.py:58]: done

File: predictions/panns/panns_inference.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+                        help="Path for csv file with predictions to be saved")
+    args = parser.parse_args()
+
+    test_df = pd.read_csv(args.path_to_input)
+    model_details = MODEL_DETAILS
+    panns_model = get_model(PANNsDense121Att, model_details["config"], model_details["weights_path"])
+
+    pred_df = predict(test_df, panns_model, model_details)
+    print("Saving preds....")
+    pred_df.to_csv(args.path_to_save_preds)
```
Comment (nandan-sridhar) [inline predictions/panns/panns_inference.py:188]:
```suggestion
    pred_df.to_csv(args.csv_path_to_save_preds)
```

File: models/panns_finetune.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+        label = torch.Tensor(zs)
+        return tensors, label
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Fintuner for PANNs model")
+    parser.add_argument("-train", "--train_path", help="Path to train dataframe (csv file)")
+    parser.add_argument("-test", "--test_path", help="Path to test dataframe (csv file)")
+    parser.add_argument("-p2p", "--path_to_prediction", help="Path to save predictions (csv file)")
+    parser.add_argument("-cfg", "-path2config", help="Path to the json config file")
```
Comment (nandan-sridhar) [inline models/panns_finetune.py:90]:
```suggestion
    parser.add_argument("-train", "--train_path", required=True, help="Path to train dataframe (csv file)")
    parser.add_argument("-test", "--test_path", required=True, help="Path to test dataframe (csv file)")
    parser.add_argument("-p2p", "--path_to_prediction", required=True, help="Path to save predictions (csv file)")
    parser.add_argument("-cfg", "-path2config", required=True, help="Path to the json config file")
```

File: model_configs/panns/train_config.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
+import torch.nn as nn
```
Comment (nandan-sridhar) [inline model_configs/panns/train_config.py:1]: This file can be removed

File: model_configs/panns/test_config.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
+from externals.models import PANNsDense121Att
```
Comment (nandan-sridhar) [inline model_configs/panns/test_config.py:1]: Instead of having this file, can we just use the train_config.json? Just pass the train_config.json as an argument to the panns_inference script.

File: models/README.md [hunk 11]
```diff
@@ -87,3 +87,20 @@ d. batch_size: Batch size<br>
...
+
+
+#### 6. Panns Finetuning File
+
+- The file finetunes the Panns model based on input data
+- Make sure to run it only if you have a GPU Machine
+- Additional files required to for testing
+    - `checked_df_train.csv`: Csv of training audio clip names
+    - `checked_df_test.csv` : Csv of testing audio clip names
+    - `final_5fold_sed_dense121_nomix_fold0_checkpoint_50_score0.7219`: PANNs weights
```
Comment (nandan-sridhar) [inline models/README.md:99]:
```suggestion
    - Initial PANNs weights (e.g. `final_5fold_sed_dense121_nomix_fold0_checkpoint_50_score0.7219`)
```

File: models/README.md [hunk 12]
```diff
@@ -87,3 +87,20 @@ d. batch_size: Batch size<br>
...
+    - `XC_Sounds_Split/`: Folder containing split wavfiles of corresponding xenocanto audioclips
+- If using custom data make sure to bring your data files to these formats
+
+- To run Panns finetune: run `python models/panns_xeno_finetune.py` with the 
+appropriate arguments from the help menu.
+- Make sure to adjust the relevant filepaths and parameters in the `model_configs/panns/train_config.py`
```
Comment (nandan-sridhar) [inline models/README.md:105]:
```suggestion
- To run Panns finetune: run `python models/panns_finetune.py` with the 
appropriate arguments from the help menu.
- Make sure to adjust the relevant filepaths and parameters in the `model_configs/panns/train_config.json`
```

File: predictions/panns/validate_preds.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    for p, t in zip(predictions, targets):
+        if t in p:
+            acc += 1
+    # print("Top 5 Accuracy: ",acc/len(predictions))
+    return acc / len(predictions)
+
+
+for bird_name in bird_names:
+    req = preds[preds["target"] == bird_name]
+    # accuracy = np.sum(req["preds"]==req["target"].values)/len(req)
```
Comment (nandan-sridhar) [inline predictions/panns/validate_preds.py:26]:
```suggestion
```

File: predictions/panns/validate_preds.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+def top5_acc(df):
+    predictions = df.preds.values
+    targets = df.target.values
+
+    acc = 0.0
+    for p, t in zip(predictions, targets):
+        if t in p:
+            acc += 1
+    # print("Top 5 Accuracy: ",acc/len(predictions))
```
Comment (nandan-sridhar) [inline predictions/panns/validate_preds.py:20]:
```suggestion
```

File: predictions/panns/panns_inference.py [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+
+        pred_df, clip_codes = prediction_for_clip(clip, model, threshold,
+                                                  clip_threshold)
+        if len(clip_codes) == 0:
+            clip_codes = ["NOCALL"]
+        if labels[ii] in clip_codes:
+            acc_counter += 1
+        if (ii + 1) % 10 == 0:
+            pbar.set_postfix(accuracy=acc_counter / total)
+        #    print("Accuracy after ",(ii+1),"Iterations is ",(acc_counter/total))
```
Comment (nandan-sridhar) [inline predictions/panns/panns_inference.py:155]:
```suggestion
```

File: predictions/panns/panns_inference.py [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+    predictions = {}
+    predictions["filepath"] = []
+    predictions["ebird_code"] = []
+    acc_counter = 0
+    total = 0
+    pbar = trange(len(paths))
+    pbar.set_description("Prediction Bar")
+    for ii in pbar:
+        time.sleep(0.01)
+        # print("Iteration ",ii+1,"of ",len(paths))
```
Comment (nandan-sridhar) [inline predictions/panns/panns_inference.py:138]:
```suggestion
```

File: models/panns_finetune.py [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+        pred_df = {}
+        pred_df["preds"] = val_preds
+        pred_df["target"] = val_targets
+        pd.DataFrame(pred_df).to_csv(args.path_to_prediction)
+        # print("Val Accuracy: ",val_acc/len(val_dataloader))
+        master_bar.set_postfix(train_loss=loss.item(), val_loss=val_loss.item(),
+                               train_acc=acc_score, val_acc=val_acc)
+
+    # TESTING
+    # print("Testing....")
```
Comment (nandan-sridhar) [inline models/panns_finetune.py:195]:
```suggestion
```

File: models/panns_finetune.py [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+
+        label = self.csv["ebird_code"][idx]
+        zs = np.zeros((1, self.num_classes))
+        zs[:, self.enc_dict[label]] = 1
+        clip, _ = librosa.load(path, sr=SR, mono=True, res_type="kaiser_fast")
+        array = clip.astype(np.float32)
+        if len(array) < 960000:
+            array = np.pad(array, (0, 960000 - len(array)), mode='constant', constant_values=0)
+        # print(path)
+        # print(len(array))
```
Comment (nandan-sridhar) [inline models/panns_finetune.py:78]:
```suggestion
```

File: predictions/panns/panns_infer_birds.py [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+        pred_df, clip_codes = prediction_for_clip(clip, model, threshold,
+                                                  clip_threshold)
+        pred_df.to_csv("Framewise_outputs.csv")
+        if len(clip_codes) == 0:
+            clip_codes = ["NOCALL"]
+        if labels[ii] in clip_codes:
+            acc_counter += 1
+        if (ii + 1) % 10 == 0:
+            pbar.set_postfix(accuracy=acc_counter / total)
+        #    print("Accuracy after ",(ii+1),"Iterations is ",(acc_counter/total))
```
Comment (nandan-sridhar) [inline predictions/panns/panns_infer_birds.py:156]:
```suggestion
```

File: predictions/panns/panns_infer_birds.py [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+    predictions = {}
+    predictions["filepath"] = []
+    predictions["ebird_code"] = []
+    acc_counter = 0
+    total = 0
+    pbar = trange(len(paths))
+    pbar.set_description("Prediction Bar")
+    for ii in pbar:
+        time.sleep(0.01)
+        # print("Iteration ",ii+1,"of ",len(paths))
```
Comment (nandan-sridhar) [inline predictions/panns/panns_infer_birds.py:138]:
```suggestion
```
